### ISSUE TITLE
[Talk] Activities and Fragments are controllers

### ISSUE THREAD

**Issue by Cartman34** [futurice/android-best-practices#89]
Hi,

I am a web MVC programmer and I developed several projects with Android and Android Studio, i consider the basic architecture as not MVC since Controllers and View are mixed. Separating layout in another file is not enough to make it MVC compliant, because you need some code processing this layout.
This is why for some needs, i created another concept, the LayoutView, this is specific to one layout and allow developer to interact between the controller (Activity/Fragment) and the Template (layout.xml). The layout's LayoutView inherits from an interface (provided by the controller).

This allow multiple layout for one controller depending on usage and allowing reusage through your different apps.
Here is an example for a login form, the known layout are commonly, basic or featured:

```
public interface LayoutView {

    public int getLayoutID();

}

public interface LoginView extends com.another.package.views.LayoutView {
    public void reset();

    public void setUp();

    Credentials getLoginForm();

    public void onLoginStart();

    public void onLoginEnd();

    public void onLoginSuccess();

    public void onLoginError();
}
```

getLayoutID() returns the layout ID \o/
setUp() is called by onCreate() or onCreateView()
reset() is called by onResume()
getLoginForm() is called when clicking Login button, this method extract Credentials from form.
Other method provide a way to show the state of the request to the user (here a Rest API request with Retrofit).

The architecture is
- controllers
  - activities
  - fragments
  - adapters
- views
  - LoginView
  - BasicLoginView
  - FeaturedLoginView
- models
  - User

This system is not yet mature.

What do you think about it ?

**Comment by jonatasleon** [comment]
Simple and interesting. Do you have a repo with a sample of this architecture?

**Comment by Cartman34** [comment]
Nope, sorry, i haven't any public repo to show this architecture, i am just starting with this architecture and i think it needs more experience before being mature and a public tool. ;-)
This enables a true separation of rendering and functional code of the application.

**Comment by jonatasleon** [comment]
You could to create one, whereas you and community would work in order for mature it.

**Comment by jonatasleon** [comment]
You're the boss. But a repo would be helpful to:
 - You have help in order to mature it;
 - Dissemination
;-)

**Comment by Cartman34** [comment]
Okay okay, maybe you should see that
https://github.com/Sowapps/android-mvc
I am open to suggestions

**Comment by peter-tackage** [comment]
We've found that separation of Android platform dependencies from business logic greatly improves testability, so your pattern certainly improves on the default Android solution in that regard. The MVP pattern is very popular in the Android dev community at the moment, but we've found that MVVM is often more flexible. See this [repo]( https://github.com/florina-muntenescu/DroidconMVVM) for an explanation and example code.
